tests/src/analysis/testqgsrastercalculator.cpp

```diff
@@ -876,6 +881,215 @@ void TestQgsRasterCalculator::testStatistics()
+
+  // append the two options
+  args.append( new QgsRasterCalcNode( 100.0 ) );
+  args.append( new QgsRasterCalcNode( 10.0 ) );
+
+  //name of the function
+  QString name( "if" );
+
+  QgsRasterCalcNode node( name, args );
+
+  QgsRasterMatrix result( 1, 1, nullptr, -9999 );
+
+  QVERIFY( node.calculate( rasterData, result ) );
+  QCOMPARE( result.data()[0], 100 );
+  QCOMPARE( result.data()[1], 10 );
+  QCOMPARE( result.data()[2], 100 );
+  QCOMPARE( result.data()[3], result.nodataValue() );
+  QCOMPARE( result.data()[4], 100 );
+  QCOMPARE( result.data()[5], 10 );
+
+}
+
+void TestQgsRasterCalculator::testFunctTypeSecondCase()
+{
+  // Test with one raster as condition, one raster first option and number as second option
+  //if (raster1 > 5.0, 100.0, 10.0)
+  QVector < QgsRasterCalcNode * > args;
+
+  // create and append the raster used in the condition
+  QgsRasterBlock m1( Qgis::DataType::Float32, 2, 3 );
+  m1.setNoDataValue( -1.0 );
+  m1.setValue( 0, 0, 15.0 );
+  m1.setValue( 0, 1, 1.5 );
+  m1.setValue( 1, 0, 13.0 );
+  m1.setValue( 1, 1, -1.0 ); //nodata
+  m1.setValue( 2, 0, 11.0 );
+  m1.setValue( 2, 1, 1.0 );
+  QMap<QString, QgsRasterBlock *> rasterData;
+  rasterData.insert( QStringLiteral( "raster1" ), &m1 );
+  args.append( new QgsRasterCalcNode( QgsRasterCalcNode::opGT, new QgsRasterCalcNode( QStringLiteral( "raster1" ) ), new QgsRasterCalcNode( 5.0 ) ) );
+
+  // create and append the raster used as first option
+  QgsRasterBlock m2( Qgis::DataType::Float32, 2, 3 );
+  m2.setNoDataValue( -2.0 );
+  m2.setValue( 0, 0, 150 );
+  m2.setValue( 0, 1, 150 );
+  m2.setValue( 1, 0, 125 );
+  m2.setValue( 1, 1, -2 ); //nodata
+  m2.setValue( 2, 0, 125 );
+  m2.setValue( 2, 1, 125 );
+  rasterData.insert( QStringLiteral( "raster2" ), &m2 );
+  args.append( new QgsRasterCalcNode( QStringLiteral( "raster2" ) ) );
+
+  // append the number used as second option
+  args.append( new QgsRasterCalcNode( 10.0 ) );
+
+  //name of the function
+  QString name( "if" );
+
+  QgsRasterCalcNode node( name, args );
+
+  QgsRasterMatrix result( 1, 1, nullptr, -9999 );
+
+  QVERIFY( node.calculate( rasterData, result ) );
+
+  QCOMPARE( result.data()[0], 150 );
+  QCOMPARE( result.data()[1], 10 );
+  QCOMPARE( result.data()[2], 125 );
+  QCOMPARE( result.data()[3], result.nodataValue() );
+  QCOMPARE( result.data()[4], 125 );
+  QCOMPARE( result.data()[5], 10 );
+}
+
+void TestQgsRasterCalculator::testFunctTypeThirdCase()
+{
+  // Test with one raster as condition, one raster first option and another as second option
+  //if (raster1 > 5.0, 100.0, 10.0)
+  QVector < QgsRasterCalcNode * > args;
+
+  // create and append the raster used in the condition
+  QgsRasterBlock m1( Qgis::DataType::Float32, 2, 3 );
+  m1.setNoDataValue( -1.0 );
+  m1.setValue( 0, 0, 15.0 );
+  m1.setValue( 0, 1, 1.5 );
+  m1.setValue( 1, 0, 13.0 );
+  m1.setValue( 1, 1, -1.0 ); //nodata
+  m1.setValue( 2, 0, 11.0 );
+  m1.setValue( 2, 1, 1.0 );
```

Comment:
if you use the same test raster block multiple times, it would be good to move the raster block initialization to a dedicated function in the test (to avoid code duplicity)

Reply:
Ok, I see what you mean. Taking into account the last two comments I decided to chage the test method a little bit. I'll skip the creation of the raster and use existing raster data (such as `landsat.tif`), if this last path is not clear or is not more understandable, I'll get back to the first approach..
I also add more complex arguments  as options.